File: src/main/java/com/hoi4utils/clausewitz/exceptions/NodeValueTypeException.java
package com.hoi4utils.clausewitz.exceptions;

import com.hoi4utils.clausewitz_parser.Node;

public class NodeValueTypeException extends Exception {
    public NodeValueTypeException(Node exp,  Class<?> clazz) {
        super("[" + clazz.getName() + "] Invalid node value type: " + exp.name());
    }

    public NodeValueTypeException(Node exp, Throwable cause, Class<?> clazz) {
        super("[" + clazz.getName() + "] Invalid node value type: " + exp.name(), cause);
    }

    public NodeValueTypeException(Node expression, String expected, Class<?> clazz) {
        super("[" + clazz.getName() + "] Invalid node value type of expression: "
                + expression.name() + ", Expected: " + expected);
    }
}

File: src/main/scala/com/hoi4utils/exceptions/NodeValueTypeException.scala
package com.hoi4utils.exceptions

import com.hoi4utils.parser.Node

/**
 * Exception thrown when a node's value type does not match the expected type.
 *
 * @param message The detail message.
 * @param cause   The cause of the exception.
 */
class NodeValueTypeException(message: String, cause: Throwable) extends Exception(message, cause):

  def this(message: String) =
    this(message, null)

  def this(exp: Node[?], clazz: Class[?]) =
    this(s"[${clazz.getName}] Invalid node value type: ${exp.name}")

  def this(exp: Node[?], cause: Throwable, clazz: Class[?]) =
    this(s"[${clazz.getName}] Invalid node value type: ${exp.name}", cause)

  def this(exp: Node[?], expected: String, clazz: Class[?]) =
    this(s"[${clazz.getName}] Invalid node value type of expression: ${exp.name}, Expected: $expected")
